File: src/features/offices/dto/create-office.dto.ts
import { Transform, Type } from 'class-transformer';
import { ArrayMaxSize, IsArray, IsInt, IsOptional, IsString } from 'class-validator';

export class CreateOfficeDto {
  @IsString()
  @Transform(({ value }) => (typeof value === 'string' ? value.trim() : value))
  name: string;

  @IsOptional()
  @IsString()
  orgId?: string;

  @IsOptional()
  @IsString()
  @Transform(({ value }) => (typeof value === 'string' ? value.trim() : value))
  address?: string;

  @IsOptional()
  @IsString()
  @Transform(({ value }) => (typeof value === 'string' ? value.trim() : value))
  description?: string | null;

  @IsOptional()
  @IsString()
  @Transform(({ value }) => (typeof value === 'string' ? value.trim() : value))
  timezone?: string | null;

  @IsOptional()
  @IsString()
  @Transform(({ value }) => (typeof value === 'string' ? value.trim() : value))
  orgLocationTypeKey?: string | null;

  @IsOptional()
  @IsArray()
  @ArrayMaxSize(100)
  @IsString({ each: true })
  @Transform(({ value }) =>
    Array.isArray(value) ? value.map((v) => (typeof v === 'string' ? v.trim() : v)) : value
  )
  tagKeys?: string[];

  @IsOptional()
  @IsString()
  @Transform(({ value }) => (typeof value === 'string' ? value.trim() : value))
  parentOrgLocationId?: string | null;

  @IsOptional()
  @IsInt()
  @Type(() => Number)
  sortOrder?: number | null;
}

File: spring-boot-api/src/main/java/com/mytegroup/api/dto/offices/CreateOfficeDto.java
package com.mytegroup.api.dto.offices;

import jakarta.validation.constraints.Max;
import jakarta.validation.constraints.Min;
import jakarta.validation.constraints.NotBlank;
import jakarta.validation.constraints.Size;

import java.util.List;

public record CreateOfficeDto(
    @NotBlank(message = "Name is required")
    String name,
    String orgId,
    String address,
    String description,
    String timezone,
    String orgLocationTypeKey,
    @Size(max = 100, message = "Tag keys must not exceed 100 items")
    List<String> tagKeys,
    String parentOrgLocationId,
    @Min(value = 0, message = "Sort order must be non-negative")
    Integer sortOrder
) {
    public CreateOfficeDto {
        if (name != null) {
            name = name.trim();
        }
        if (address != null) {
            address = address.trim();
        }
        if (description != null) {
            description = description.trim();
        }
        if (timezone != null) {
            timezone = timezone.trim();
        }
        if (orgLocationTypeKey != null) {
            orgLocationTypeKey = orgLocationTypeKey.trim();
        }
        if (parentOrgLocationId != null) {
            parentOrgLocationId = parentOrgLocationId.trim();
        }
    }
}


